### ISSUE TITLE
Ingress Controller Scalability Optimizations

### ISSUE THREAD

**Issue by learnitall** [cilium/cilium#28699]
There are a few scalability optimizations that can be made in the Ingress Controller to reduce load on the K8s API server in larger clusters.

* **Duplicate Informers**: The Ingress Controller has its own Informers for Services and Endpoints, however, the Operator has `Resource[T]` based Informers for Services and Endpoints in place. Migrating the Ingress Controller to use these shared Informers would reduce the load on the K8s API server by removing the duplicate watches.
* **Inefficient logic to regenerate shared CiliumEnvoyConfig**: When the Ingress Controller is acting in shared mode, anytime an Ingress resource is updated the entire shared CiliumEnvoyConfig is regenerated and updated. This may add additional load on the K8s API server, as well as additional resource usage in the Cilium Operator, during the startup of the Ingress Controller while the Ingress Controller initially processes each Ingress resource.

Data would need to be collected in order to understand how much of an impact these optimizations would actually have.

**Comment by prateek041** [comment]
Hey @learnitall 

I have worked with the tetragon project but fairly new to cilium's github. I would like to work on this issue since it seems to be an improvement (research) based issue with no rigid deadlines. 

Reading from the description, I understand that Ingress Controller is producing load on the k8s API for larger clusters, and we can prevent that from happening by using same (shared) informers instead of different ones, also the logic of generating `ciliumEnvoyConfig` can be improved.

Would you like to guide me on this issue ? would be great help.

**Comment by learnitall** [comment]
Hey @prateek041, go for it! I can help guide you on this issue. Thank you for volunteering to pick this up. What can I do to help?

**Comment by prateek041** [comment]
Thank you ! @learnitall 

For the starters, Just point me to the relevant files, so I can get better understanding of how the current code is handling things.

**Comment by prateek041** [comment]
After going through the issue mentioned by @youngnick above, I came to this conclusion, please correct if I am wrong.

- Currently the [Ingress controller](https://github.com/cilium/cilium/blob/main/operator/pkg/ingress/ingress.go) here is using Informers.
- [Gateway API](https://github.com/cilium/cilium/blob/main/operator/pkg/gateway-api/controller.go) on the other hand is using Controller runtime.

We have to migrate the logic to Controller runtime for Ingress too.

cc @learnitall

**Comment by learnitall** [comment]
Hey @prateek041, since @mhofstetter is working on the above CFP, which should address this issue anyway, I don't think it makes sense to do any work for this issue individually at this time. Let's let @mhofstetter take this one. If there is another issue that catches your eye, please feel free to pick it up.

**Comment by prateek041** [comment]
ok, thanks for letting me know.

**Comment by youngnick** [comment]
Thanks for that @learnitall, and sorry, I've been sick the last couple of days.

@prateek041, thanks for offering to look into this, I'm going to leave this open and we'll see if any further changes are required once #28911 is done.
